### PR TITLE
[VPlan] Support VPReverseVectorPointer in DataWithEVL vectorization

### PR DESCRIPTION
**Pull request by arcbbb**
VPReverseVectorPointer relies on the runtime VF, but in DataWithEVL tail-folding, EVL (which can be less than VF at runtime) should be used instead. 

This patch updates the logic to check the users of VF and replaces the second operand if the user is VPReverseVectorPointer.

### CODE REVIEW

**Comment by arcbbb** [comment]
rebase after #110974 is merged.

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp
Comment (arcbbb) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp]: Nice catch. I think this can occur when V is loop-invariant. Fixed.

Comment (arcbbb) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp]: Fixed. Thanks!

**Comment by arcbbb** [comment]
ping

**Review by Mel-Chen — APPROVED** [review]
LGTM, but wait a few days to see if other reviewers have any opinions. 
I think merging it in next week should be better since today is Friday.:)

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:1475]: Would it be better to set EVL for all `VPReverseVectorPointerRecipe`? There could be other memory recipes in the future , e.g. interleave groups

File: llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:162]: Can be handled with `.Case<VPWidenLoadEVLRecipe>(`?

**Comment by arcbbb** [comment]
rebase to solve conflict.

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp
Comment (arcbbb) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp]: Updated, instead I traverse each user of VF to find all `VPReverseVectorPointerRecipe`.

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:1452]:
```suggestion
    if (auto *R = dyn_cast<VPReverseVectorPointerRecipe>(U))
```

**Review by fhahn — APPROVED** [review]
LGTM, with suggestion inline

**Pull request merged**
